### DISCUSSION TITLE
Can not get the dll injected into any game

### DISCUSSION THREAD

**Discussion by Godnoken** [veeenu/hudhook#33] [General]
Hello hello!

As referenced in #32, I can not successfully inject the dll file into any game that I've tried so far.

I start the game and use the generated .exe - a cmd window pops up for a split second and disappears again.
I've tried launching the game like in the README.md, but to no avail - and some games refuse to start that way due to Steam, Epic Games etc.
I have no antivirus installed and I turned off the windows defender & firewall.
OS is Windows 10.
Games tried are Overgrowth, Kenshi, Bloons TD 6, SHENZHEN IO, Fallout 3.

Any idea what I may be doing wrong?

**lib.rs**

Exactly as the OpenGL3 hook [example](https://github.com/veeenu/hudhook/blob/master/examples/opengl3_hook.rs)

**main.rs**
````
use hudhook::inject;
use std::process::Command;

fn main() {
     test_run_against_sample();
}


fn test_run_against_sample() {

    inject::inject("Overgrowth.exe", "target/debug/my_lib.dll".into()).ok();
}
````

**Cargo.toml**
```

[dependencies]
hudhook = "0.2.0"
imgui = "0.8.0"

[lib]
crate-type = ["cdylib"]
name = "my_lib"
```

**Comment by veeenu** [comment]
I don't know any of those games' internals, but I googled Fallout 3 and it looks like it is a DirectX 9 game.

You should choose the hook that works with the rendering API that each game is using. 

For example, Dark Souls III uses DirectX 11 and Elden Ring uses DirectX 12, so in the respective tools I use `ImguiDX11Hooks` and `ImguiDX12Hooks` in the generic at the `hudhook!` macro call site.

**Reply by Godnoken** [reply]
Sorry for not being clear, but yes, I have done that. The code example above was just for Overgrowth specifically.
So I'm not doing anything wrong code-wise as far as you can see?

I'll try to install a few more games tomorrow and see if I can get it to work.

**Reply by veeenu** [reply]
- Could you upload your projects' code up somewhere, so I can take a look at it? The culprit might be something small that's not easy to see at first blush.
- Does the game crash as well, or does it continue running?
- If the former, could you try and attach a debugger to the running game to see what is causing the crash?

**Reply by Godnoken** [reply]
> Could you upload your projects' code up somewhere, so I can take a look at it? The culprit might be something small that's not easy to see at first blush.

https://github.com/Godnoken/hudhook

It's set up for GTA: San Andreas now as it is supposed to work according to one of the issues in this repo.
By the way - I'm not supposed to have to add imgui-rs as a dependency myself, am I? For some reason it doesn't acknowledge that it is included in the hudhook crate.

> Does the game crash as well, or does it continue running?

Nope, nothing happens apart from a cmd window popping up for half a second. Game keeps running.
I've even looked into everything happening with the game's .exe with [Process Explorer](https://learn.microsoft.com/en-us/sysinternals/downloads/process-explorer) and I can't see anything indicating that hudhook is doing anything to it.

**Reply by veeenu** [reply]
Mmm, can you try supplying an absolute path to the DLL? It's counterintuitive (and should be documented), but the injection works by calling `LoadLibrary` in the context of the .exe, so the DLL path that gets passed as string should be relative to that. Or, preferably, absolute, which you can build at runtime off of a relative path; this is by far the most robust solution. See [this](https://github.com/veeenu/eldenring-practice-tool/blob/main/practice-tool/src/main.rs#L64) for an example.

**Reply by Godnoken** [reply]
Did not make a difference unfortunately..

Do you think it may have anything to do with the games being launched through Steam/Epic games?

**Reply by Godnoken** [reply]
Could you send me your toolchain setup from the ``rustup show`` command?

**Reply by Godnoken** [reply]
Got it working with 64-bit applications, didn't realize that 32-bit isn't fully supported.. Should have read the GTA issue thread more thoroughly heh.

Although it is interesting that he got GTA: SA working whereas I can not. Not sure if my x86 build isn't working properly.

Do you know what needs to be done to adopt the hooks for 32-bit?

**Reply by veeenu** [reply]
Sorry for the late replies, have been afk!

I don't have `rustup show`'s output handy but iirc I'm on the nightly from about a week ago. Practice tools' CI pipelines use the latest though and it still works as I've published releases a couple days ago.

I never really considered 32-bit due to always working with 64-bit code for my use cases -- I think @FrankvdStam has been dealing with that for Dark Souls PTDE and may have more insights in that, but unless all the other games you mentioned are 32-bit I think that may not be the problem.

Long shot: do you have non-ASCII characters in the path to your DLL? i.e. umlauts in the username or something like that. It is an old issue that should have been _long gone_ but we recently switched to `dll-syringe` for the injection backend and who knows, it may have cropped back up.

**Reply by Godnoken** [reply]
No worries mate!

> but unless all the other games you mentioned are 32-bit I think that may not be the problem.

Some of the games - BloonsTD6, Kenshi & Overgrowth were indeed not 32-bit. The issue with the hudhook injector was that I wrote in the application name together with the .exe extension. Leaving it out makes it possible to inject to BloonsTD6 and Overgrowth, but not Kenshi. However, I tried this CLI [Injector](https://github.com/nefarius/Injector) and it successfully injects into Kenshi.
I also noticed that the CLI Injector seemingly has no issues with any antivirus (at least not Bitdefender & Windows), while hudhooks injector does. Not sure if it's anything you can affect? Nefarius mentions that his Injector is likely deemed safe due to being around for a few years at https://github.com/nefarius/Injector/issues/15

> Long shot: do you have non-ASCII characters in the path to your DLL?

Nope, nothing like that.

**Reply by Godnoken** [reply]
I previously thought that my 32-bit dll didn't work with the CLI Injector either - turns out I was wrong.
I naively used the 64-bit CLI injector and thought that it was backwards compatible. With the 32-bit version I can now inject the DX9 hook into any 32-bit instance, just not with your dll-syringe.

Is dll-syringe set up for x64 only in this crate perhaps? I saw that the dll-syringe docs say that using the into-x86-from-x64 feature would make 32-bit injection possible.

**Reply by Godnoken** [reply]
I can confirm that dll-syringe works for x86 with the ``into-x86-from-x64`` feature after some testing.

However,

> we recently switched to dll-syringe for the injection backend 

I'm sorry for my ignorance, but I can not see that you're using it anywhere in this library. Am I misunderstanding something?

**Reply by veeenu** [reply]
> I'm sorry for my ignorance, but I can not see that you're using it anywhere in this library. Am I misunderstanding something?

I was conflating things -- I use `dll-syringe` in the practice tools in lieu of `inject::inject` as it is generally more complete than my implementation. I'm planning to either deprecate that method, or make it a frontend to `dll-syringe`. Haven't really been focusing on that front lately as there are countless methods of injecting DLLs whereas the problem of building a DLL and hooking functions is a bit more complex to tackle.

**Reply by Godnoken** [reply]
Ahh, gotcha.

Yeah, completely understood now, although it wasn't for a newbie like me who's never delved into these things before.
Integrating dll-syringe into this project would be neat!

Thank you for the help 🥇

**Reply by veeenu** [reply]
Oh well, don't worry -- it's tricky every time even for hardened veterans 😂 

I need to do a round-up on the injection techniques out there, see if there are notable differences, but making `inject::inject` a `dll-syringe` frontend would definitely solve most issues without having to reinvent the wheel I think.

**Comment by FrankvdStam** [comment]
![image](https://user-images.githubusercontent.com/37239092/197392655-63de387b-1b21-4815-9cc4-c5c7d33d0474.png)

32bit is working just fine, you're likely not injecting correctly. You can use something like cheat engine to see if your dll is actually loaded inside the target process. Also to make things easier on yourself, allocate a console from inside the target process so that you can use println! or othe macro's to get some logging, and redirect panics so that you can actually see if something inside your code is crashing. 

```Rust
    util::console::init_console();  //calls AllocConsole() from windows-rs
    util::log::init_log();          //Initializes a console logger with log4rs (can just use println! instead after AllocConsole())

    //Redirect panics, log them to console and to file with log4rs, once again use println! as alternative - you won't see 
    //crashes if you don't log to file because the console will close.
    panic::set_hook(Box::new(|i| {
        error!("panic");
        error!("{}", i);
    }));
```


There are definitely bugs left in the dx9 backend. The most glaring one being that the mouse position you see on screen does not match up with where imgui thinks the mouse is.

**Reply by Godnoken** [reply]
Thank you for the input Frank!

I've used Process Explorer to see if the DLL gets injected & it does not.

Eitherway, forgive me if I am wrong - but are you even using hudhook::inject in your soulmemory-rs repo? Because I can inject 32-bit just fine if I use other injectors, but not that one.

I also noticed the mouse bug. It's accurate sometimes but it seems that if you move fast it breaks.

**Reply by FrankvdStam** [reply]
I use my own lib for the injection https://github.com/FrankvdStam/mem-rs/blob/60ebb93e812f42860d885ba1017c4f787bb5b974/src/process.rs#L103

I assume they work very similar though. The most important part, as veeenu mentioned, is getting the filepath right, and to make sure it's null terminated like in C, when handing it to win32's LoadLibraryA. Different libraries will handle that differently I assume. mem-rs will null-terminate the string for you. I don't know the implementation details of how hudhook does the injection. mem-rs is very new though, and I'm probably the sole user, you're better off using a popular library I guess.

During testing you can always just rely on cheat engine for the injection. If your injection doesn't work, but cheat engine's does, then your code is broken somewhere.

As for the mouse issue, I don't think it matters how fast you move the mouse, I think it has to do with how far the mouse has moved from the top left corner. I think that the mouse moves 2 pixels in the imgui layer for every 1 pixel it moves on screen. 

Also has to do with there being multiple framebuffers in dx9 with different sizes. Needs more investigation to fully understand, its fairly complicated.

**Reply by Godnoken** [reply]
Had no clue about the null termination but it didn't seem to make a difference in ``inject::inject``.
Eitherway, I ended up using dll-syringe and it seemingly works very well.

Yeah, I'm sure you have an understanding of it that is a lot better than mine - however in my tests, sometimes it did register the mouse position correctly and not only in the top left corner. Seems like something triggers it or maybe resets at some point?

**Reply by veeenu** [reply]
I have never written DX9 code myself, but it does feel like a scaling issue which may or may not crop up depending on the situation. Like all graphics things, it's super hard to debug.

**Reply by Godnoken** [reply]
Interestingly, for whatever it is worth - the mouse pos seems to have no problem tracking accurately in Morrowind & San Andreas on DX9 but not in Fallout 3.

**Comment by veeenu** [comment]
@Godnoken I have opened [a branch](https://github.com/veeenu/hudhook/compare/andrea/inject-improvements?expand=1) for improving the general behavior of `inject` in lieu of using `dll-syringe`, as I decided I'd rather have fewer dependencies.

I also added some tests to inject an example dummy DLL into Notepad. To run these tests:

```
> cargo build --examples
> cargo test -- --nocapture
```

Could you please try the injector in that branch against the games you are working with?

I think I didn't specify this earlier because I'm a dummy, and this could very well be a solution -- you have to specify the _title_ to the window rather than the name of the process, though this could (and probably should?) change in the future. Example with `notepad`:

```rust
let mut child = Command::new("notepad.exe").spawn().expect("Couldn't start notepad");
std::thread::sleep(Duration::from_millis(500)); // wait for the process to start

hudhook::inject::inject("Untitled - Notepad", examples_path().join("dummy_hook.dll")).unwrap();
```

I'm planning to introduce a breaking change in the API for `0.3.0` which will have one `inject(exe_name, dll_path)` and one `inject_by_title(title, dll_path)` or something to that tune.

Thank you!

**Reply by Godnoken** [reply]
Awesome bud!

I can confirm that the injector from the new branch works on all the games I tried.
No issues with Windows internal antivirus blocking it either. Don't have any other antivirus to test it against at the moment however.

Having the possibility of a choice between title & process & maybe even PID (?) would indeed be great.

---

Sorry for the lack of activity - I have relocated to a new country & had other things to do other than programming (unfortunately :p)

**Reply by veeenu** [reply]
That's great to hear! I'm tracking this effort on https://github.com/veeenu/hudhook/issues/35, landing in the next release 😁

No worries about lack of activity! I see any amount of time anyone spends on open source at all as a gift, so, rather, thank you for taking the time to check the new feats! And gl with the relocation, these things usually take a lot of effort even when it just sounds like "I went from point A to point B". 😄

**Reply by Godnoken** [reply]
Awesome work mate. I tried the new injection overhaul too and it works like a charm.

Haha, I'm glad you see it that way - I just wish I could help out a lot more. From the research I made, this seems to be exactly what I need for my app to reach its full potential. So I will stick around & hopefully sooner rather than later, I can contribute code too.

Thank you! Yeah, it's definitely not just A to B, but god do I wish it was... This is the 5th country that I've lived in. No idea why I put myself through it over and over :p

**Reply by veeenu** [reply]
>Haha, I'm glad you see it that way - I just wish I could help out a lot more.

I'm happy with any and all contributions, at any pace at all, so no worries! The project is going to stick around whenever you want 😄 

>Yeah, it's definitely not just A to B, but god do I wish it was... This is the 5th country that I've lived in. No idea why I put myself through it over and over :p

Well, on the plus side, hopefully you're seeing a lot of cool stuff of the world! I've only ever left my country for travel, never for living abroad, and I expect that to be a completely different experience.

**Reply by Godnoken** [reply]
> Well, on the plus side, hopefully you're seeing a lot of cool stuff of the world! I've only ever left my country for travel, never for living abroad, and I expect that to be a completely different experience.

Absolutely, many fond memories. Wouldn't recommend moving often though, it's more of a pain than enjoyment hehe 😅
